src/function/scalar/date/date_part.cpp

```diff
@@ -372,485 +372,397 @@ struct DateDatePart {
-	template <class TA, class TB, class TR>
-	static inline TR Operation(TA specifier, TB date) {
-		return ExtractElement<TB, TimeDatePart>(GetDatePartSpecifier(specifier.GetString()), date);
+template <typename TA, typename TR>
+static void DatePartUnaryExecutor(DatePartSpecifier type, Vector &left, Vector &result, idx_t count) {
+	switch (type) {
+	case DatePartSpecifier::YEAR:
+		UnaryExecutor::Execute<TA, TR, DatePart::YearOperator>(left, result, count);
+		break;
+	case DatePartSpecifier::MONTH:
+		UnaryExecutor::Execute<TA, TR, DatePart::MonthOperator>(left, result, count);
+		break;
+	case DatePartSpecifier::DAY:
+		UnaryExecutor::Execute<TA, TR, DatePart::DayOperator>(left, result, count);
+		break;
+	case DatePartSpecifier::DOW:
+		UnaryExecutor::Execute<TA, TR, DatePart::DayOfWeekOperator>(left, result, count);
+		break;
+	case DatePartSpecifier::ISODOW:
+		UnaryExecutor::Execute<TA, TR, DatePart::ISODayOfWeekOperator>(left, result, count);
+		break;
+	case DatePartSpecifier::DOY:
+		UnaryExecutor::Execute<TA, TR, DatePart::DayOfYearOperator>(left, result, count);
+		break;
+	case DatePartSpecifier::DECADE:
+		UnaryExecutor::Execute<TA, TR, DatePart::DecadeOperator>(left, result, count);
+		break;
+	case DatePartSpecifier::CENTURY:
+		UnaryExecutor::Execute<TA, TR, DatePart::CenturyOperator>(left, result, count);
+		break;
+	case DatePartSpecifier::MILLENNIUM:
+		UnaryExecutor::Execute<TA, TR, DatePart::MilleniumOperator>(left, result, count);
+		break;
+	case DatePartSpecifier::QUARTER:
+		UnaryExecutor::Execute<TA, TR, DatePart::QuarterOperator>(left, result, count);
+		break;
+	case DatePartSpecifier::WEEK:
+		UnaryExecutor::Execute<TA, TR, DatePart::WeekOperator>(left, result, count);
+		break;
+	case DatePartSpecifier::EPOCH:
+		UnaryExecutor::Execute<TA, TR, DatePart::EpochOperator>(left, result, count);
+		break;
+	case DatePartSpecifier::MICROSECONDS:
+		UnaryExecutor::Execute<TA, TR, DatePart::MicrosecondsOperator>(left, result, count);
+		break;
+	case DatePartSpecifier::MILLISECONDS:
+		UnaryExecutor::Execute<TA, TR, DatePart::MillisecondsOperator>(left, result, count);
+		break;
+	case DatePartSpecifier::SECOND:
+		UnaryExecutor::Execute<TA, TR, DatePart::SecondsOperator>(left, result, count);
+		break;
+	case DatePartSpecifier::MINUTE:
+		UnaryExecutor::Execute<TA, TR, DatePart::MinutesOperator>(left, result, count);
+		break;
+	case DatePartSpecifier::HOUR:
+		UnaryExecutor::Execute<TA, TR, DatePart::HoursOperator>(left, result, count);
+		break;
+	default:
+		throw NotImplementedException("Specifier type not implemented for DATEPART");
 	}
-};
+}
+
+template <typename T>
+static void DatePartFunction(DataChunk &args, ExpressionState &state, Vector &result) {
+	D_ASSERT(args.ColumnCount() == 2);
+	auto &part_arg = args.data[0];
+	auto &date_arg = args.data[1];
+
+	if (part_arg.GetVectorType() == VectorType::CONSTANT_VECTOR) {
+		// Common case of constant part.
+		if (ConstantVector::IsNull(part_arg)) {
```

Comment:
For `date_part` this case is already handled through the `DatePartSimplificationRule` rule, which converts date_part with a constant into individual functions (e.g.  `EXTRACT('YEAR' FROM time)` becomes `YEAR(time)`). Perhaps not worth the added code size of having this case handled here separately as well?

Reply:
Oh right. There were other things worth keeping in here, though, including the general cleanup to match the other "part" function internals and adding support for `TIME` types. I've also updated the documentation to include all the missing functions and types.